File: idl-conformance/_interop/ts/src/xcdr1_decode.ts
// XCDR1 DECODE conformance: decode the ZeroDDS-Rust XCDR1 (classic CDR /
// PL_CDR1) reference goldens with the TS binding via decode(.., representation=0)
// and assert every field == canonical. Proves the TS decoder reads the second
// wire representation (no DHEADER on @appendable/@final, 8-byte align, PL_CDR1
// for @mutable) — the counterpart of be_roundtrip.ts.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { feat } from "../gen/30_features.js";
const T = feat as any;

const HERE = dirname(fileURLToPath(import.meta.url));
const GOLD = join(HERE, "..", "..", "..", "proofs", "endianness", "goldens");

const samples: Array<[string, any, any]> = [
  ["wstr", T.WStrTypeSupport, { label: "café", text: "日本語\u{1F389}" }],
  ["mut",  T.MutTypeSupport,  { a: 1_000_000, b: 2.5, c: "ok" }],
  ["mapenum", T.MapEnumTypeSupport, { h: T.Hue.H_BLUE, m: new Map([[3, {x:11,y:12}]]), sels: [{discriminator:2, n:9}] }],
  ["tree", T.TreeTypeSupport, { value:1, kids:[{value:2,kids:[{value:4,kids:[]}]},{value:3,kids:[]}] }],
  ["arr",  T.ArrTypeSupport,  { grid:[[10,11,12],[13,14,15]], shape:[{x:1,y:2},{x:3,y:4}] }],
  ["prim", T.PrimTypeSupport, undefined],
];

const norm = (v: any) =>
  JSON.stringify(v, (_k, x) => (x instanceof Map ? [...x.entries()] : (typeof x === "bigint" ? x.toString() : x)));

let fail = 0;
for (const [name, ts, sample] of samples) {
  let bytes: Uint8Array;
  try {
    bytes = new Uint8Array(readFileSync(join(GOLD, `${name}.xcdr1-le.rust.bin`)));
  } catch {
    console.log(`  ${name} xcdr1: SKIP (no golden)`);
    continue;
  }
  // DECODE: representation = 0 selects the XCDR1 decode path.
  const back = ts.decode(bytes, 0, bytes.length, "le", 0);
  // `prim` has no inline canonical here — just assert it decodes without throwing
  // and round-trips a couple of distinctive fields.
  const ok = sample === undefined
    ? (back && typeof back === "object")
    : norm(back) === norm(sample);
  console.log(`  ${name} xcdr1 decode: ${ok ? "OK" : "FAIL"}`);
  if (!ok) { fail = 1; console.log(`    want ${norm(sample)}\n    got  ${norm(back)}`); }
  // ENCODE: re-encode the decoded value (no inline canonical needed) and assert
  // byte-identity to the golden.
  const wire: Uint8Array = ts.encode(back, "le", 0);
  const encOk = wire.length === bytes.length && wire.every((b: number, i: number) => b === bytes[i]);
  console.log(`  ${name} xcdr1 encode: ${encOk ? "OK" : "FAIL"} (${wire.length}B vs ${bytes.length}B)`);
  if (!encOk) fail = 1;
}
process.exit(fail);

File: idl-conformance/_interop/python/xcdr1_decode.py
"""XCDR1 DECODE conformance: decode the ZeroDDS-Rust XCDR1 (classic CDR /
PL_CDR1) reference goldens with the Python binding and assert every field
equals the canonical sample. Proves the Python decoder reads the XCDR1 wire
(no DHEADER on @appendable/@final, max-align 8, PL_CDR1 for @mutable) — the
counterpart of be_roundtrip.py for the second wire representation."""
import os, sys
_HERE = os.path.dirname(os.path.abspath(__file__))
_REPO = os.path.abspath(os.path.join(_HERE, "..", "..", "..", ".."))
sys.path.insert(0, os.path.join(_REPO, "crates", "py", "python"))
sys.path.insert(0, _HERE)
import features_interop as fi

_GOLDENS = os.path.join(_HERE, "..", "..", "proofs", "endianness", "goldens")

# Every feature that has an XCDR1-LE golden (arr/bits/mapenum/mut/prim/tree/wstr).
cases = [
    ("wstr", fi.canonical_wstr()),    # @appendable
    ("mut", fi.canonical_mut()),      # @mutable -> PL_CDR1
    ("mapenum", fi.canonical_mapenum()),
    ("tree", fi.canonical_tree()),
    ("arr", fi.canonical_arr()),
    ("prim", fi.canonical_prim()),    # @appendable + 8-byte members (align 8)
]

fail = 0
for name, sample in cases:
    path = os.path.join(_GOLDENS, f"{name}.xcdr1-le.rust.bin")
    if not os.path.exists(path):
        print(f"  {name} xcdr1: SKIP (no golden)")
        continue
    with open(path, "rb") as f:
        data = f.read()
    cls = type(sample)
    # DECODE: representation=0 selects the XCDR1 decode path.
    back = cls.decode(data, "le", 0)
    ok = fi.equal(back, sample)
    print(f"  {name} xcdr1 decode: {'OK' if ok else 'FAIL'}")
    if not ok:
        fail = 1
        print(f"    want {sample}\n    got  {back}")
    # ENCODE: encode(.., representation=0) must be byte-identical to the golden.
    wire = sample.encode("le", 0)
    enc_ok = wire == data
    print(f"  {name} xcdr1 encode: {'OK' if enc_ok else 'FAIL'}"
          f" ({len(wire)}B vs {len(data)}B golden)")
    if not enc_ok:
        fail = 1

sys.exit(fail)
